static/app/components/events/interfaces/exception.tsx

```diff
@@ -10,88 +10,83 @@ import {ExceptionType} from 'app/types';
 import {Event} from 'app/types/event';
 import {STACK_TYPE, STACK_VIEW} from 'app/types/stacktrace';
 
-const defaultProps = {
-  hideGuide: false,
-};
-
 type Props = {
   event: Event;
   type: string;
   data: ExceptionType;
   projectId: string;
-} & typeof defaultProps;
-
-type State = {
-  stackView: STACK_VIEW;
-  stackType: STACK_TYPE;
-  newestFirst: boolean;
+  hideGuide?: boolean;
 };
 
-class Exception extends Component<Props, State> {
-  static defaultProps = {
-    hideGuide: false,
-  };
+function Exception({event, type, data, projectId, hideGuide = false}: Props) {
+  const [stackView, setStackView] = useState<STACK_VIEW>(
+    data.hasSystemFrames ? STACK_VIEW.APP : STACK_VIEW.FULL
+  );
+  const [stackType, setStackType] = useState<STACK_TYPE>(STACK_TYPE.ORIGINAL);
+  const [newestFirst, setNewestFirst] = useState(isStacktraceNewestFirst());
 
-  state: State = {
-    stackView: this.props.data.hasSystemFrames ? STACK_VIEW.APP : STACK_VIEW.FULL,
-    newestFirst: isStacktraceNewestFirst(),
-    stackType: STACK_TYPE.ORIGINAL,
-  };
+  const eventHasThreads = !!event.entries.find(entry => entry.type === 'threads');
 
-  handleChange = (newState: Partial<State>) => {
-    this.setState(prevState => ({
-      ...prevState,
-      ...newState,
-    }));
-  };
+  // in case there are threads in the event data, we don't render the
+  // exception block.  Instead the exception is contained within the
+  // thread interface.
+  if (eventHasThreads) {
+    return null;
+  }
 
-  render() {
-    const eventHasThreads = !!this.props.event.entries.find(
-      entry => entry.type === 'threads'
-    );
+  function handleChange({
+    stackView: newStackView,
+    stackType: newStackType,
+    newestFirst: newNewestFirst,
+  }: {
+    stackView?: STACK_VIEW;
+    stackType?: STACK_TYPE;
+    newestFirst?: boolean;
+  }) {
+    if (newStackView && newStackView !== stackView) {
+      setStackView(newStackView);
+    }
 
-    // in case there are threads in the event data, we don't render the
-    // exception block.  Instead the exception is contained within the
-    // thread interface.
-    if (eventHasThreads) {
-      return null;
+    if (newNewestFirst && newNewestFirst !== newestFirst) {
+      setNewestFirst(newNewestFirst);
     }
 
-    const {projectId, event, data, hideGuide, type} = this.props;
-    const {stackView, stackType, newestFirst} = this.state;
+    if (newStackType && newStackType !== stackType) {
+      setStackType(newStackType);
+    }
```

Comment:
Just wondering, do we need all these `!==` checks here? Will the setState be intelligent enough to not cause the rerender if you set the same value?

Reply:
will not cause rerender... thanks 🙂